src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs

```diff
@@ -23,11 +53,81 @@ public string V3Search(V3SearchRequest request)
             return GetLuceneQuery(request.Query);
         }
 
-        private static string GetLuceneQuery(string query)
+        private string GetLuceneQuery(string query)
         {
-            // TODO: query parsing
-            // https://github.com/NuGet/NuGetGallery/issues/6456
-            return query ?? "*";
+            if (string.IsNullOrWhiteSpace(query))
+            {
+                return MatchAllDocumentsQuery;
+            }
+
+            var grouping = _parser.ParseQuery(query.Trim());
+            if (!grouping.Any())
+            {
+                return MatchAllDocumentsQuery;
+            }
+
+            var result = ToAzureSearchQuery(grouping).ToString();
+            if (string.IsNullOrWhiteSpace(result))
+            {
+                return MatchAllDocumentsQuery;
+            }
+
+            return result;
+        }
+
+        private AzureSearchQuery ToAzureSearchQuery(Dictionary<QueryField, HashSet<string>> grouping)
+        {
+            var result = new AzureSearchQuery();
+
+            foreach (var field in grouping)
+            {
+                // Add values that aren't scoped to a field.
+                if (field.Key == QueryField.Any)
+                {
+                    result.AddNonFieldScopedValues(field.Value);
+                }
+                else if (field.Key != QueryField.Invalid)
+                {
+                    // Add values that are scoped to a valid field.
+                    var fieldName = FieldNames[field.Key];
+                    var values = ProcessFieldValues(field.Key, field.Value);
+
+                    result.AddFieldScopedValues(fieldName, values);
+                }
+             }
+
+            return result;
+        }
+
+        private static IEnumerable<string> ProcessFieldValues(QueryField field, IEnumerable<string> values)
+        {
+            switch (field)
+            {
+                // The "packageId" and "owners" fields are case insensitive exact matches
+                case QueryField.PackageId:
+                case QueryField.Owner:
+                    return values.Select(v => v.ToLowerInvariant());
```

Comment:
My understanding is that you should not need to lowercase a search term because the analyzer associated with the search field will be used. See `analyzer` vs. `searchAnalyzer` vs. `indexAnalyzer`:
https://docs.microsoft.com/en-us/azure/search/search-what-is-an-index#components-of-an-index

My understanding is that we are using `analyzer` which uses the same analyzer for both searching and indexing.

Reply:
[The analyzer for exact matches](http://lucene.apache.org/core/7_3_1/analyzers-common/org/apache/lucene/analysis/core/KeywordAnalyzer.html) is case sensitive. To work around this, you can either:

1. Use [a filter to lowercase terms](https://lucene.apache.org/core/4_10_3/analyzers-common/org/apache/lucene/analysis/core/LowerCaseFilter.html)
2. Lowercase the field's content in the index and the search query.

The first approach would require building a custom Azure Search Analyzer that chains the lowercase filter and the keyword analyzer. Both approaches results in a new field that is no longer displayable as the content has lost its casing. The second approach is easier to implement, which is why used that here. Does that make sense?

Reply:
I've reverted my changes for case insensitive exact match semantics, this will re-added by https://github.com/NuGet/NuGetGallery/issues/6920